src/fiber/fiber.ml

```diff
@@ -394,6 +394,52 @@ module Ivar = struct
       | Empty _ -> None )
 end
 
+module Mvar = struct
+  type 'a t =
+    { writers : ('a * unit Ivar.t) Queue.t
+    ; readers : 'a Ivar.t Queue.t
+    ; mutable value : 'a option
+    }
+
+  let create () =
+    { value = None; writers = Queue.create (); readers = Queue.create () }
+
+  let next_writer (type a) (t : a t) =
+    if Queue.is_empty t.writers then
+      None
+    else
+      let a, w = Queue.pop t.writers in
+      assert (t.value = None);
+      t.value <- Some a;
+      Some w
+
+  let read (type a) (t : a t) k =
+    match t.value with
+    | None ->
+      let ivar = Ivar.create () in
+      Queue.add ivar t.readers;
+      Ivar.read ivar k
+    | Some v -> (
+      t.value <- None;
+      match next_writer t with
+      | None -> k v
+      | Some w -> Ivar.fill w () (fun () -> k v) )
```

Comment:
@jeremiedimino I'm not sure I like this behavior. It seems like `read` should be called immediately with `v` and we shouldn't wait for another writer to succeed.

Reply:
Nvm, this is no longer a problem once i removed the ivars